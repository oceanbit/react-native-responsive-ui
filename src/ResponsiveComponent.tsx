import { Component } from "react";
import { Dimensions, ScaledSize } from "react-native";

interface ResponsiveComponentProps {}

interface ResponsiveComponentState {
  window: ScaledSize;
}

export default abstract class ResponsiveComponent extends Component<
  ResponsiveComponentProps,
  ResponsiveComponentState
> {
  state = {
    // eslint-disable-next-line react/no-unused-state
    window: Dimensions.get("window")
  };

  componentWillMount() {
    Dimensions.addEventListener("change", this.onDimensionChange);
  }

  componentWillUnmount() {
    Dimensions.removeEventListener("change", this.onDimensionChange);
  }

  onDimensionChange = (dims: ResponsiveComponentState) => this.setState(dims);
}
