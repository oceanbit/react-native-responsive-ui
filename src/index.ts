export { default as useDimensions } from "./useDimensions";
export { default as useStylesheet } from "./useStylesheet";
export { default as mediaQuery, IMediaQuery } from "./MediaQuery";
