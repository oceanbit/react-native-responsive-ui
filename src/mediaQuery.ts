import { Platform, PixelRatio } from "react-native";

type Orientation = "landscape" | "portrait";

export interface MediaQuery {
  minHeight?: number;
  maxHeight?: number;
  minWidth?: number;
  maxWidth?: number;
  minAspectRatio?: number;
  maxAspectRatio?: number;
  minPixelRatio?: number;
  maxPixelRatio?: number;
  orientation?: Orientation;
  condition?: boolean;
  platform?: string;
}

const isInInterval = (value: number, min?: number, max?: number): boolean =>
  (min === undefined || value >= min) && (max === undefined || value <= max);

const mediaQuery = (
  {
    minWidth,
    maxWidth,
    minHeight,
    maxHeight,
    minAspectRatio,
    maxAspectRatio,
    orientation,
    platform,
    minPixelRatio,
    maxPixelRatio,
    condition
  }: MediaQuery,
  width: number,
  height: number
): boolean => {
  const currentOrientation: Orientation =
    width > height ? "landscape" : "portrait";
  return (
    isInInterval(width, minWidth, maxWidth) &&
    isInInterval(height, minHeight, maxHeight) &&
    isInInterval(width / height, minAspectRatio, maxAspectRatio) &&
    isInInterval(PixelRatio.get(), minPixelRatio, maxPixelRatio) &&
    (orientation === undefined || orientation === currentOrientation) &&
    (platform === undefined || platform === Platform.OS) &&
    (condition === undefined || condition)
  );
};

export default mediaQuery;
