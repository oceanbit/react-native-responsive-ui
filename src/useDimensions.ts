import { useEffect, useState } from "react";
import { Dimensions, ScaledSize } from "react-native";

const dims = Dimensions.get("window");

export default () => {
  const [dimensions, setDimensions] = useState(dims);

  const onChange = ({
    window: { width, height, scale, fontScale }
  }: {
    window: ScaledSize;
  }) => {
    setDimensions({ width, height, scale, fontScale });
  };

  useEffect(() => {
    Dimensions.addEventListener("change", onChange);
    return () => Dimensions.removeEventListener("change", onChange);
  }, []);

  return dimensions;
};
