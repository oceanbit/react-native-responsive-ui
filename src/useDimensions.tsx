import { useEffect, useState } from "react";
import { Dimensions, ScaledSize } from "react-native";

const dims = Dimensions.get("window");

export default () => {
  const [dimensions, setDimensions] = useState(dims);

  const onChange = ({ window }: { window: ScaledSize }) =>
    setDimensions(window);

  useEffect(() => {
    Dimensions.addEventListener("change", onChange);
    return () => Dimensions.removeEventListener("change", onChange);
  }, []);

  return dimensions;
};
