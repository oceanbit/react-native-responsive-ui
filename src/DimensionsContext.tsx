import React from "react";
import { Dimensions } from "react-native";

import useDimensions from "./useDimensions";

export const DimensionsContext = React.createContext(Dimensions.get("window"));

interface DimensionProviderProps {
  children: React.ReactNode;
}

export const DimensionsConsumer = DimensionsContext.Consumer;

export const DimensionsProvider = ({ children }: DimensionProviderProps) => {
  const dimensions = useDimensions();
  return (
    <DimensionsContext.Provider value={dimensions}>
      {children}
    </DimensionsContext.Provider>
  );
};
