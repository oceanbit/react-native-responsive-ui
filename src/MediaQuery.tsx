import React from "react";
import { PixelRatio, Platform } from "react-native";
import useDimensions from "./useDimensions";

type Orientation = "landscape" | "portrait";

export interface IMediaQuery {
  minHeight?: number;
  maxHeight?: number;
  minWidth?: number;
  maxWidth?: number;
  minAspectRatio?: number;
  maxAspectRatio?: number;
  minPixelRatio?: number;
  maxPixelRatio?: number;
  orientation?: Orientation;
  condition?: boolean;
  platform?: string;
}

export const isInInterval = (
  value: number,
  min?: number,
  max?: number
): boolean =>
  (min === undefined || value >= min) && (max === undefined || value <= max);

export const mediaQuery = (
  {
    minWidth,
    maxWidth,
    minHeight,
    maxHeight,
    minAspectRatio,
    maxAspectRatio,
    orientation,
    platform,
    minPixelRatio,
    maxPixelRatio,
    condition
  }: IMediaQuery,
  width: number,
  height: number
): boolean => {
  const currentOrientation: Orientation =
    width > height ? "landscape" : "portrait";
  return (
    isInInterval(width, minWidth, maxWidth) &&
    isInInterval(height, minHeight, maxHeight) &&
    isInInterval(width / height, minAspectRatio, maxAspectRatio) &&
    isInInterval(PixelRatio.get(), minPixelRatio, maxPixelRatio) &&
    (orientation === undefined || orientation === currentOrientation) &&
    (platform === undefined || platform === Platform.OS) &&
    (condition === undefined || condition)
  );
};

interface MediaQueryProps extends IMediaQuery {
  children?: React.ReactNode;
}

export default ({ children, ...props }: MediaQueryProps) => {
  const { width, height } = useDimensions();
  const val = mediaQuery(props, width, height);
  if (val) {
    return children;
  }
  return null;
};
