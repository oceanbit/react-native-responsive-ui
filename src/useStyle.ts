import { ScaledSize, ViewStyle, TextStyle, ImageStyle } from "react-native";
import { mediaQuery, IMediaQuery } from "./MediaQuery";
import useDimensions from "./useDimensions";

type Style = ViewStyle | TextStyle | ImageStyle;

let id = 0;
const uniqueId = () => {
  id += 1;
  return id;
};

type ReponsiveStyles = Style & {
  [condition: string]: Style;
};

const selectStyle = (styles: ReponsiveStyles): Style => ({
  ...(Object.keys(styles)
    .filter(style => style.endsWith("-true"))
    .map(style => styles[style]) as Style)
});

export const useStyle = (
  styles: (cond: (query: IMediaQuery) => string) => ReponsiveStyles
): Style => {
  const dimensions = useDimensions();
  const cond = (query: IMediaQuery) =>
    `${uniqueId()}-${mediaQuery(dimensions, query)}`;
  return selectStyle(styles(cond));
};

export const getStyle = (
  dimensions: ScaledSize,
  styles: (cond: (query: IMediaQuery) => string) => ReponsiveStyles
): Style => {
  const cond = (query: IMediaQuery) =>
    `${uniqueId()}-${mediaQuery(dimensions, query)}`;
  return selectStyle(styles(cond));
};
