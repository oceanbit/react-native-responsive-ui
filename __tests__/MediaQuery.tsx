import { View } from "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { MediaQuery } from "../src";

it("renders MediaQuery", () => {
    const tree = renderer
    .create(
      <View>
          <MediaQuery minHeight={450} orientation="portrait">
              test
          </MediaQuery>
      </View>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
})