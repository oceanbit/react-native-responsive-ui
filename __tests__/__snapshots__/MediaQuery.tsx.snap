// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MediaQuery 1`] = `
<View>
  test
</View>
`;
